File: bot/exceptions/validation_exception.py
"""Exception levée lorsqu'un paramètre libre (callback_data, argument de
commande, corps de webhook) ne respecte pas le format attendu.
"""

from __future__ import annotations


class ValidationException(Exception):
    def __init__(self, errors: dict[str, str], message: str = "Validation échouée.") -> None:
        super().__init__(message)
        self.errors = errors

File: bot/vicia-bot/bot/Core/Exceptions/ValidationException.php
<?php

namespace Bot\Core\Exceptions;

/**
 * Class ValidationException
 *
 * Levée lorsqu'une saisie utilisateur (texte libre envoyé en réponse
 * à une invite du bot — e-mail, code, valeur de seuil...) ne respecte
 * pas le format attendu. Porte la liste des erreurs par champ, pour
 * un éventuel réaffichage détaillé du formulaire conversationnel.
 */
class ValidationException extends BotException
{
    /** @param array<string, string> $errors champ => message */
    public function __construct(private readonly array $errors, string $userMessage = 'Saisie invalide.')
    {
        parent::__construct($userMessage, 'Validation échouée : ' . json_encode($errors, JSON_UNESCAPED_UNICODE));
    }

    public function errors(): array
    {
        return $this->errors;
    }

    public function firstError(): ?string
    {
        return array_values($this->errors)[0] ?? null;
    }
}
